Filter to columns used in downstream analysis, convert to pandas dataframe, and split chromosomes

KatalinaBobowik opened the pull request:
Filter to columns used in downstream analysis, convert to pandas dataframe, and save each chromosome as individual files. [Successfully](https://batch.hail.populationgenomics.org.au/batches/8682/jobs/2) run in `test`

Commit:
Subset all chromosomes to individual parquet files

Commit:
Add comment on what I'm doing when splitting files

Commit:
Filter columns before converting to pandas df, to save memory

Commit:
Densify mt after reading it in

Commit:
Remove all instances of 'locus.contig' in script to see if it finishes

Commit:
Mkae flatten = True explicit, to see if this solves for compresssion of pandas df

Commit:
Split columns into two

Commit:
Save intermediate file to play around with

Commit:
Add more print statements to debug failed script

Commit:
Add in another print statement to debug

Commit:
More debugging with print messages

Commit:
COnvert locus_alleles df to string

Commit:
Remove print messages and see if converting to string solves str split issue

Commit:
Update path to saved files and use output_path function

lgruen commented on scripts/generate_genotype_df/generate_genotype_df.py:
```diff
@@ -24,8 +25,15 @@ def query():
+    # select only locus and alleles, which are the keys, then convert to pandas
+    locus_alleles = mt.rows().select().to_pandas(flatten=True)
+    # expand locus to two columns and rename 
+    locus_alleles = locus_alleles.astype(str)
```
Instead of converting `Locus` to a string and then pulling it apart using `split`, can you use the `contig` and `position` fields directly?
https://hail.is/docs/0.2/genetics/hail.genetics.Locus.html#hail.genetics.Locus

MattWellie replied:
I tried to do a matrix.locus.contig.collect_as_set() but the notebook ran out of memory!

KatalinaBobowik replied:
@MattWellie I've never seen `collect_as_set()` before. I just tried it on a small subset now. 

@lgruen The issue I'll have if I use `mt.locus.contig` or `mt.locus.position` directly is that it gets converted to a string expression rather than staying as a hail table (which has the `to_pandas` functionality). The best I could come up with instead is something like this:
`mt = mt.annotate_rows(contig = mt.locus.contig, position=mt.locus.position).key_rows_by('alleles')
locus_alleles = mt.rows().select('contig','position').to_pandas(flatten=True)
`

I can't select rows by `mt.locus.contig` or `mt.locus.position` directly, unfortunately

lgruen replied:
Could you instead say:
```
t = mt.rows()
t = t.key_by(contig = t.locus.contig, position = t.locus.position)
t = t.select(t.alleles)
pd = t.to_pandas(flatten=True)
```
?

KatalinaBobowik replied:
that's a lot more readable. I've incorporated this version in. Thanks!

lgruen commented on scripts/generate_genotype_df/generate_genotype_df.py:
```diff
@@ -13,9 +14,9 @@ def query():
     hl.init(default_reference='GRCh38')
 
     mt = hl.read_matrix_table(TOB_WGS)
+    mt = hl.experimental.densify(mt)
```
Do you need this? Won't this just affect the reference blocks?

KatalinaBobowik replied:
I thought densifying was necessary whenever you use the `GT` field?

lgruen replied:
Sorry, of course you're right. I keep forgetting about the actual semantics of `densify`. Loic put it pretty succinctly a while ago: "densify does not add or remove rows, ever. It simply fills up missing entries." (https://nbviewer.org/github/populationgenomics/tob-deepdive/blob/main/densify.ipynb).

Commit:
Annotate rows by locus and contig, then convert to df (rather than splitting after)

lgruen reviewed (changes requested):
Looks like there's also a merge conflict that you'll need to resolve first.

KatalinaBobowik commented:
@lgruen looking into the merge conflict now

Commit:
Update conversion of selecting chr, contigs, and alleles

This version, suggetsed by Leo, is easier to follow, and the alleles also come last (in teh correct order).

Commit:
Update comment to reflect new changes

Commit:
Merge branch 'main' into split_chromosomes

Commit:
Fix pylint errors (extra carriage return and removing spaces around '='

Commit:
Add 2 blank lines after function (as per pylint)

Commit:
Remove carriage return at end of file (pylint)

Commit:
Add new line at end of file

Commit:
Remove indent from last line
